daemon/src/engine/server/CryptoChallenge.h

```diff
@@ -0,0 +1,203 @@
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+===========================================================================
+*/
+
+#ifndef CRYPTOCHALLENGE_H
+#define CRYPTOCHALLENGE_H
+
+#include "framework/Crypto.h"
+
+class Challenge
+{
+public:
+    using Clock = Sys::SteadyClock;
+    using TimePoint = Clock::time_point;
+    using Duration = Clock::duration;
+
+    /*
+     * Time duration a challenge will be valid for
+     */
+    static Duration Timeout();
+
+    /*
+     * Size of the raw challenge data
+     */
+    static std::size_t Bytes();
+
+    Challenge( const netadr_t& source, const Crypto::Data& challenge )
+        : created( Clock::now() ),
+          timeout( Timeout() ),
+          challenge( challenge ),
+          source( source )
+    {
+        // The port number might change in connectionless commands
+        this->source.port = this->source.port4 = this->source.port6 = 0;
+    }
+
+    explicit Challenge( const netadr_t& source )
+        : Challenge( source, Crypto::RandomData( Bytes() ) )
+    {
+    }
+
+    /*
+     * Maximum time this challenge should be considered valid
+     */
+    TimePoint ValidUntil() const
+    {
+        return created + timeout;
+    }
+
+    /*
+     * Whether the challenge is valid at the given time point
+     */
+    bool ValidAt( const TimePoint& time ) const
+    {
+        return created + timeout >= time;
+    }
+
+    /*
+     * Whether two challenges match
+     */
+    bool Matches( const Challenge& other ) const
+    {
+        if ( challenge != other.challenge )
+            return false;
+
+        return NET_CompareAdr( source, other.source );
+    }
+
+    /*
+     * Challenge as a hex string
+     */
+    std::string String() const
+    {
+        return Crypto::String( Crypto::Encoding::HexEncode( challenge ) );
+    }
+
+    /*
+     * Time when this challenge has been created
+     */
+    TimePoint Created() const
+    {
+        return created;
+    }
+
+    /*
+     * Duration this challenge has been alive for
+     */
+    Duration Lifetime() const
+    {
+        return Clock::now() - created;
+    }
+
+    /*
+     * Checks whether a string is a proper challenge representation
+     * (ie: something that String() might have returned
+     */
+    static bool ValidString(const std::string& challenge);
+
+private:
+    /*
+     * Generates a random challenge
+     */
+    std::string GenerateString();
+
+    TimePoint    created;
+    Duration     timeout;
+    Crypto::Data challenge;
+    netadr_t     source;
+
+};
+
+class ChallengeManager
```

Comment:
Benefit of a singleton vs a namespace + methods? Seem like the latter will work well enough.